### PR TITLE
Change comparison for Network::Address implementations to be based on address compare, not string compare

### LINKED ISSUE #3529: Address::Instance equality implemented through string compare
**Issue Template**

*Title*: Address::Instance equality implemented through string compare

*Description*:
The currently implementations of Address::Instance (Ipv4instance, Ipv6Instance, and PipeInstance) all inherit from InstanceBase, which implements operator== as:

  bool operator==(const Instance& rhs) const override { return asString() == rhs.asString(); }

This works, but is a) dependent on the derived classes having a 1-1 mapping between string representations and unique instances, and b) not a very efficient method of computing equality for Ipv4 and Ipv6 addresses.  If we implement any algorithms that need to scale and compare Instances frequently, the string compare cost could be significant.

Unfortunately, the current method has the advantage of working (modulo (a) above) on all derived classes of Address::Instance, current&future, core envoy and downstream specializations as it is defined in terms of methods on Instance.  Individual derived types could use RTTI to determine if the type being compared to was a known type, and implement the comparison in a more efficient fashion if it was.  If it is desirable that RTTI be avoided, a visitor design pattern could be implemented on the Instance to do double dispatch on the types of the two objects being compared.   However, this involves a non-trivial change to the interface of the abstract base class.

@htuch : I hadn't figured out when we were talking what the implications of "We should do this upstream" were.  Given the above paragraph, do you have any prefereces?

### PR DESCRIPTION
**Pull request by curiouserrandy**
*title*: Change comparison for Network::Address implementations to be based on address compare, not string compare

*Risk Level*: Low

*Testing*: Existing unit tests for comparison.

*Docs Changes*: N/A

*Release Notes*: N/A

Fixes #3529

### CODE REVIEW

**Comment by curiouserrandy** [comment]
Matt: Ready for next review whenever you have a chance.

File: test/common/network/address_impl_test.cc [hunk 1]
```diff
@@ -392,6 +392,62 @@ TEST(AddressFromSockAddr, Pipe) {
 #endif
 }
 
+// Test comparisons between all the different (known) test classes.
+TEST(MixedAddresses, Equality) {
```
Comment (mattklein123) [inline test/common/network/address_impl_test.cc]: nit: Is it possible to somehow use TEST_P to do all of the mix and match logic? I'm guessing it is, but I'm not a gtest expert. Not a big deal if it's not easy.

File: test/common/network/address_impl_test.cc [hunk 2]
```diff
@@ -392,6 +392,62 @@ TEST(AddressFromSockAddr, Pipe) {
...
+  enum InstanceType { kIpv4, kIpv6, kPipe };
```
Comment (mattklein123) [inline test/common/network/address_impl_test.cc]: nit: we don't use 'k' prefix for constance. I would just drop the 'k'.

File: test/common/network/address_impl_test.cc (hunk 1 above)
Comment (curiouserrandy) [reply test/common/network/address_impl_test.cc]: Glancing at the docs for TEST_P (https://github.com/google/googletest/blob/master/googletest/docs/advanced.md, search for TEST_P) I don't see how to do this.  The problem is that we want a cross-product of all the possible addresses, and TEST_P only takes a single list of parameters.  I could make a structure that contained both things to compare, but I don't see a way to auto-generate that structure, which means I'm effectively replacing two for loops with a hardcoded list.  That doesn't strike me as a step forward for readability or maintainability.

Do you see anything I'm missing?

Comment (mattklein123) [reply test/common/network/address_impl_test.cc]: I thin you can do something with `testing::Combine` like here: https://github.com/envoyproxy/envoy/blob/master/test/common/http/http2/codec_impl_test.cc#L669 but I'm not completely sure. It's not a big deal either way.

Comment (curiouserrandy) [reply test/common/network/address_impl_test.cc]: Ah, neat.  Done.  (I'm not *sure* it was worth it, but I learned something so I'll call that a win :-}.)

File: test/common/network/address_impl_test.cc (hunk 2 above)
Comment (curiouserrandy) [reply test/common/network/address_impl_test.cc]: Done.

**Pull request merged**